cpp/oneapi/dal/backend/primitives/reduction/reduction_rm_rw.cpp

```diff
@@ -0,0 +1,393 @@
+    const BinaryOp binary,
+    const UnaryOp unary,
+    const event_vector& deps) const {
+    return this->operator()(input, output, width, height, width, binary, unary, deps);
+}
+
+template <class Float, class BinaryOp, class UnaryOp>
+sycl::nd_range<2> reduction_rm_rw_wide<Float, BinaryOp, UnaryOp>::get_range(
+    const std::int64_t height) const {
+    const sycl::range<2> local(1, wg);
+    const sycl::range<2> global(height, wg);
+    return sycl::nd_range<2>(global, local);
+}
+
+template <class Float, class BinaryOp, class UnaryOp>
+typename reduction_rm_rw_wide<Float, BinaryOp, UnaryOp>::kernel_t
+reduction_rm_rw_wide<Float, BinaryOp, UnaryOp>::get_kernel(const Float* input,
+                                                           Float* output,
+                                                           const std::int64_t width,
+                                                           const std::int64_t stride,
+                                                           const BinaryOp binary,
+                                                           const UnaryOp unary) {
+    return kernel_t(input, output, width, stride, binary, unary);
+}
+
+#define INSTANTIATE(F, B, U) template class reduction_rm_rw_wide<F, B, U>;
+
+#define INSTANTIATE_FLOAT(B, U)                \
+    INSTANTIATE(double, B<double>, U<double>); \
+    INSTANTIATE(float, B<float>, U<float>);
+
+INSTANTIATE_FLOAT(min, identity)
+INSTANTIATE_FLOAT(min, abs)
+INSTANTIATE_FLOAT(min, square)
+
+INSTANTIATE_FLOAT(max, identity)
+INSTANTIATE_FLOAT(max, abs)
+INSTANTIATE_FLOAT(max, square)
+
+INSTANTIATE_FLOAT(sum, identity)
+INSTANTIATE_FLOAT(sum, abs)
+INSTANTIATE_FLOAT(sum, square)
+
+#undef INSTANTIATE_FLOAT
+
+#undef INSTANTIATE
+
+template <class Float, class BinaryOp, class UnaryOp>
+class kernel_reduction_rm_rw_narrow {
+    typedef const Float* inp_t;
+    typedef Float* out_t;
+
+public:
+    kernel_reduction_rm_rw_narrow(inp_t const input_,
+                                  out_t const output_,
+                                  const std::int32_t width_,
+                                  const std::uint64_t height_,
+                                  const std::int32_t lstride_,
+                                  const BinaryOp binary_,
+                                  const UnaryOp unary_)
+            : input{ input_ },
+              output{ output_ },
+              unary{ unary_ },
+              binary{ binary_ },
+              width{ width_ },
+              height{ height_ },
+              lstride{ lstride_ } {}
+    void operator()(sycl::nd_item<1> it) const {
+        // Common for whole WG
+        const auto row_idx = it.get_global_id(0);
+        if (row_idx < height) {
+            // It should be conmverted to upper type by default
+            inp_t const inp_row = input + lstride * row_idx;
+            // Exclusive for EU
+            Float acc = binary.init_value;
+            for (std::int32_t i = 0; i < width; ++i) {
```

Comment:
It is reading with a stride equal to width

Reply:
We define stride separately to be able to reduce submatrix

Reply:
Discussed